### PR TITLE
Add warnings (and annotations) to PromQL query results

### PR DESCRIPTION
**Pull request by zenador**
Addresses some of https://github.com/prometheus/prometheus/issues/10839 and https://github.com/prometheus/prometheus/issues/11216 as a hackathon project.

Most of the code changes are just changing from `storage.Warnings` to `notes.Warnings` to move it to a more neutral location as the new warnings we are adding are not related to storage.

Annotations are created and handled separately from warnings here, but they are merged together with the warnings in the PromQL engine because the HTTP API and the UI don't support it yet and we want to avoid making external changes for now. By creating annotations separately though, it would make it easier in future when we want to display them differently.

Due to the way warnings are created, we end up with many duplicate warnings, or perhaps we find the warning is not needed later on, e.g. for the `RangeTooSmallWarning`, we may add this warning at each section of the range, or we may only add it at the beginning or end but still return a non-empty result for the query. For now we do the deduplication and the removal of the unnecessary warning at the final response writing step in the API.

New warnings (`util/notes/warnings.go`):
- RangeTooSmallWarning         = errors.New("Need at least 2 points to compute, perhaps time range is too small")
- MixedFloatsHistogramsWarning = errors.New("Range contains a mix of histograms and floats")
- MixedOldNewHistogramsWarning = errors.New("Range contains a mix of conventional and native histograms")
- InvalidQuantileWarning    = errors.New("Quantile value should be between 0 and 1")
- BadBucketLabelWarning     = errors.New("No bucket label or malformed label value")
- PossibleNonCounterWarning = errors.New("Metric might not be a counter (name does not end in _total/_sum/_count)")

Possible future improvements:
- We might want to change `notes.Warnings` ([]error) in the PromQL engine to structs that contain both warnings and annotations (the new `notes.Notes`) to delay the merging of warnings and annotations (which would be helpful if we eventually separate them in the final output), and to have it as an independent struct to make any future changes easier (not done yet as it is not strictly necessary at this stage and I'm not sure if we want to do this)
- With the above change, we can do the deduplication at an earlier stage each time we merge Notes to avoid collecting a long list of duplicate warnings until the final step. With a struct, we can easily add internal maps to support the deduplication in the merge function.

### CODE REVIEW

File: util/notes/warnings.go
Comment (beorn7) [inline util/notes/warnings.go:43]: I see what you are trying here, but I'm afraid it won't really solve the problem:
* We will still see false positives if the selected vector happens to have only one time series, and that series legitimately has just started (so that the range may be ever so long, it would not see another sample).
* This will definitely create false negatives, for example if there are many series in the vector and the user has picked a range close to the scrape interval. In that case, some series might succeed to yield a rate, while most won't.

File: util/notes/warnings.go
Comment (beorn7) [inline util/notes/warnings.go:33]: I see three issues here:
1. The easy one: It should be "short", not "small", shouldn't it?
2. The moderately hard one: I think this should not be a warning, but an annotations (or info, or whatever nomenclature we end up with). I would say, a warning is when something definitely went wrong, but there are reasons to not fail the entire query because if it. However, it can be totally legitimate to have only 1 sample within the range (most prominently if the series has just started or stopped).
3. The really really hard nut to crack: This is just super prone to false positives (and even the idea with only reporting it for completely empty results doesn't help that much – but it makes things more complicated because now we can also have false negatives). As long as we only add this as an annotation, we could deem the false positives as acceptable. However, I think it will create more confusion than help.

So what would be the "right" conditions to issue this annotation, beyond having only one or even just zero samples in the range? My improvised attempt: All of the following must be true:

1. There must be no staleness marker in the range (because that means we are looking at the end of a series, which would explain that we don't see at least two samples).
2. There must be samples within a reasonable time _before_ the range so that there is even a chance to find more samples by extending the range. (This is hard in two ways: First, what is a reasonable time to look back? Currently, Prometheus doesn't know the intended scrape interval of a series. Maybe the lookback delta would be good as a start. Second, this requires a separate query to find out, which is a completely new flow.)
3. The last sample before the range must not be a staleness marker (because that hints towards an actual gap in the series, which is a case where it's expected to see only one sample in a range).

File: util/notes/warnings.go
Comment (beorn7) [inline util/notes/warnings.go:50]: Would be good to also include the affected metric.

File: util/notes/warnings.go
Comment (beorn7) [inline util/notes/warnings.go:39]: We generally start error strings with lower case.

Comment (beorn7) [inline util/notes/warnings.go:39]: This should also be an annotation/info-level (because the naming is just a recommendation, and the query might be legit, see thoughts above about the range warning).

**Review by beorn7 — CHANGES_REQUESTED** [review]
Thanks for doing this. As you can see from the comments, I'm not sure yet how we should tackle this problem. Comments and thoughts from @codesome, @roidelapluie, or any other interested maintainer are highly welcome.

As a minor point: Apparently, the linter has given up on enforcing a doc comment for every exported type. However, for many of the new exported types in this PR, a doc comment would be really helpful. Could you go through them and add a doc comment wherever it appears even remotely plausible to do so?

**Comment by beorn7** [comment]
So, in summary, I think we have to crack to nuts here:
1. How to deal with the "not enough samples in the range" warning/annotation.
2. How to implement the plumbing for conveying warnings/annotations at all.

To reduce the problem surface, I think we should focus on (2) for now and only start thinking about (1) once the plumbing is in place.

WRT (2): To give some preference among the many thoughts I have expressed in the comments, my gut feeling is currently to…
1. …not think about deduplication but go for more specific warnings/annotations so that deduplication might not be needed (but could be revisited if needed).
2. …treat everything as Go `error`s utilizing the [errors package tooling](https://pkg.go.dev/errors) (which you have already done partially anyway).

In that way, we won't need any structures (like `notes.Notes`) but just specific error values and types. Maybe have an error value called `PromQLAnnotation` and then an (unexported) type `promQLAnnotation` with an `Is` method returning true for `target == PromQLAnnotation`.

File: web/api/v1/api.go
Comment (zenador) [reply web/api/v1/api.go:1588]: I think we would still need deduplication as the exact same warning would still appear multiple times for the same metric. You're right that it's not so ideal to do it here. What do you think of my suggestions under `Possible future improvements` in the PR description?

File: util/notes/warnings.go
Comment (zenador) [reply util/notes/warnings.go]: You mean move `PossibleNonCounterWarning` to `annotations.go` and add it in the code as an annotation?

File: util/notes/warnings.go
Comment (zenador) [reply util/notes/warnings.go]: It depends on how accurate we want this particular warning to be. If we just want to suggest a possible reason to the user why the query returns an empty result when doing a `rate` or something, and don't mind false positives/negatives (we could make it clear in the text that it might be wrong), then it's fine, but otherwise it seems like we'd need some complicated logic or to run additional queries just for a warning. In any case, I agree that we should just drop this for now, focus on adding the plumbing for warnings and the easy warnings first, and then tackle this in a future PR, so I've removed this warning and the associated logic from this one.

File: web/api/v1/api.go
Comment (beorn7) [reply web/api/v1/api.go:1626]: If there aren't that many duplicates, it might even be OK to leave them in. But given that there might be a _lot_ of duplicates, it's even a resource usage problem to drag those many warnings through the whole stack, only to deduplicate them here, in the last step, taking a lot of resources again. I would rather go for a deduplication on the fly from the beginning to avoid that. (Note that this code lands in the main prometheus codebase and will affect many many users, of which some might get bitten hard if their Prometheus server suddenly takes some percent more RAM. And I would like to avoid hiding the enhanced warnings behind yet another feature flag.)

File: util/notes/warnings.go
Comment (beorn7) [reply util/notes/warnings.go]: Yeah, kind of, but I'm not sure yet if we should have a distinction between `Info` and `Warning` at all (or at least not that fundamentally as currently in the code).

File: util/notes/warnings.go
Comment (beorn7) [reply util/notes/warnings.go]: OK, so let's postpone the range-to-short discussion for later. (Maybe somebody has found a proper solution for the "xrate" problem by then, and we don't have users anymore that accidentally pick a too short range…)

**Review by beorn7 — COMMENTED** [review]
My apologies again for taking so long. This is a big deal, both in reward and effort, and smaller (also important) things kept preempting this bulky thing.

Thank you very much for addressing a lot of the detailed comments (and postponing the "range too short" issue). Before diving into a detailed code level review, let's find a way forward with the fundamental structure of the annotations. See my comments here. I'll also try to write down a proposal in the next hour or so. Maybe we should jump on a call to discuss our options in detail.

**Comment by beorn7** [comment]
Here some rough ideas how to approach the problem. Happy to discuss here or in a call to refine.

- Let's make the `annotation.Annotations` type the only thing we hand around and propagate (essentially replacing `storage.Warnings` in the current upstream code). And let's get rid of `annotations.Info` and `annotations.Errors`.
- `annotations.Annotations` could have `map[string]error` as the underlying type, with the stringified version of each `error` as the key, thereby providing de-duplication.
- The `error`s we put into that map could wrap marker errors in a suitable way (see [Go documentation](https://pkg.go.dev/errors#Is) and details below).

To expand on the last point:

Here is a slight variation of what you had already done:

```Go
var (
	PromQLInfo    = errors.New("PromQL info")
	PromQLWarning = errors.New("PromQL warning")

	BadBucketLabelWarning  = fmt.Errorf("%w: no bucket label or malformed label value", PromQLWarning)
	PossibleNonCounterInfo = fmt.Errorf("%w: metric might not be a counter (name does not end in _total/_sum/_count)", PromQLInfo)
)

func NewBadBucketLabelWarning(metricName, label string) error {
	return fmt.Errorf("%w: %s %s", BadBucketLabelWarning, metricName, label)
}

func NewPossibleNonCounterInfo(metricName string) error {
	return fmt.Errorf("%w: %s", PossibleNonCounterInfo, metricName)
}
```

For this example, I have only implemented two annotations here: `BadBucketLabelWarning` precisely as before, and `NewPossibleNonCounterInfo` as an info version of the previous `PossibleNonCounterWarning`.

Now a new version of the API could return the info annotations in a separate section by iterating through all annotations and testing them with `errors.Is(annotation, annotations.PromQLInfo)` and putting them into the new section if true or in the old warnings section if false, more or less like this:

```Go
for _, annotation := range annotations {
    if errors.Is(annotation, annotations.PromQLInfo) {
        // Add to info section.
    } else {
        // Add to warning section as before.
    }
}
```

**Comment by zenador** [comment]
@beorn7 I will fix the test errors soon (but the build works, you can test the warnings with `quantile(3, up)`), but meanwhile please have a look to see if this is what you meant.

File: util/annotations/annotations.go
Comment (beorn7) [inline util/annotations/annotations.go:22]: I wouldn't explicitly call out Grafana here. Maybe be more API centric, along the lines: "... for warnings and other information that is returned by the query API along with the results."

This should also explain that an individual annotation is modeled by a Go `error`, and that `Annotations` deduplicates based on the string returned by `error.Error()`.

File: util/annotations/annotations.go
Comment (beorn7) [inline util/annotations/annotations.go:29]: I would call this `New` and without the `error` parameter (creating an empty `Annotations`). Combined with `Add` above, you can then simply write `a := annotations.New().Add(err)`

With the proposed safeguard for `Add` (see above), you could also not have this method at all and write `annotations.Annotations{}.Add(err)`, but that has stutter again, so maybe a `New` function is better.

File: util/annotations/annotations.go
Comment (beorn7) [inline util/annotations/annotations.go:44]: I would just use combinations of methods with short names rather than having this explicit method, e.g. `annotations.New().Add(err).Merge(aa)`.

In any case, I think the cases where this method is used could be avoided anyway, see comment there.

File: util/annotations/annotations.go
Comment (beorn7) [inline util/annotations/annotations.go:78]: conventional → classic (that's the canonical name we have agreed upon).

File: storage/secondary.go
Comment (beorn7) [inline storage/secondary.go:54]: I don't think you have to create a new `Annotations` here. Wouldn't the following work?
```Go
return nil, w.Add(err), nil
```

File: util/annotations/annotations.go
Comment (beorn7) [inline util/annotations/annotations.go:61]: The keys in the maps are already the strings created by `err.Error()`. You can use them directly.

File: util/annotations/annotations.go
Comment (beorn7) [inline util/annotations/annotations.go:25]: I would call this just `Add`. Idiomatic Go tries to avoid this kind of stutter.
And I would also put the safeguard in here is an `Merge` to create the map if `*a` is still `nil`.

File: util/annotations/annotations.go
Comment (beorn7) [inline util/annotations/annotations.go:37]: `*a = Annotations{}` also works (and we generally prefer that form in the Prometheus codebase).

**Review by beorn7 — CHANGES_REQUESTED** [review]
I just had a quick look at the new `Annotations`. I think that's generally the right direction. I left same nitpicky Go style comments, though.

I'm about to leave for my vacation, so I'm afraid we can finish this up only after my return.
The most important thing is to make a call how to change the API. I see the following options:

1. Simply add everything to the warnings section (I believe that's the current implementation).
2. Same as (1) but behind a feature flag.
3. Create a new info section for the info-level annotations and add warning-level annotations to the existing warnings.
4. Same as (3) but behind a feature flag.

I think (1) is a bit problematic, especially if we later decide to have an info section after all (breaking change!). Using a feature flag is the more conservative approach in any case. I'm not sure what our stability guarantees are about new sections in the JSON response. If that's OK, I would go for (3). We have to ask the experts for the API to decide.

**Comment by beorn7** [comment]
I'm back from vacation, and I'll review this ASAP.

Have you found out anything about the Prometheus stability guarantees for the API? Can we just add another key to the JSON? Otherwise, we should inquire via the developer community channels.

**Review by beorn7 — CHANGES_REQUESTED** [review]
I think this looks quite good now. I have already left a few detailed comments.

I see two substantial things left to do before we can merge this

1. How to add the annotations to the API output and whether to hide them behind a feature flag for now. @roidelapluie I would like your input so that we can make a call on this, see comment below where I mentioned you.
2. Adding a position range to the annotations where possible. I'll add my thoughts about how to do this in a separate comment.

**Comment by beorn7** [comment]
Here my thoughts about adding position info:

[Here](https://github.com/prometheus/prometheus/blob/d73b4acb301df244eacb0d50189096abc177bd07/promql/parser/parse.go#L122) is the the definition of a `ParseErr`. It contains a field of type `parser.PositionRange` and the whole `Query` as a string. It also has logic in its `Error() string` method to use this data to format a string with the correct position information (line number and position in line). I don't think we can just embed `ParseErr` into the errors in the `annotations` package, but maybe extract the logic somewhere where it can be used by both.

The more problematic part is how to even get the `PositionRange` and the full query as a string. I think the former mostly works. For example, if a PromQL function acts on a `parser.VectorSelector`, it can simply access the `PosRange` field in the `VectorSelector`. However, I haven't seen any obvious way to access the full query as a string. Maybe the engine could enrich the annotations with that string as the last step before handing back the query result including annotations to the caller.

**Comment by zenador** [comment]
@beorn7 Updated according to your comments, but still not sure how to add position info. For example, look at how we add the warning for invalid quantile in https://github.com/prometheus/prometheus/blob/f22466d167f28e3fac5ab9bc95a9c42bc7ed8fd2/promql/functions.go#L642 , I'm not sure how to get anything relevant from vals, args or enh there. We could manually pass in the string "quantile_over_time" if that helps though, and pass along the original function string for helper functions like `histogramRate`.

**Comment by beorn7** [comment]
In this case, you could use the `args` arg.

Let's assume `annotations.NewInvalidQuantileWarning` would have a `parser.PositionRange` argument, then you could write:
```Go
annos.Add(annotations.NewInvalidQuantileWarning(q, args[0].PositionRange()))
```

That's the easy part, as said. The hard part is how to give the annotation a pointer to the full query string (needed to render the error string). We could try to set that later (once the evaluation is done), but we have to be careful to not interfere with the deduplication. Or we give the `EvalNodeHelper` a pointer to the query string…

If this turns out to be too involved, we could also implement it in a later PR. I already like things how they look now:

image

**Comment by beorn7** [comment]
Wild thought: Maybe we could store the query string once per `Annotations` (make it a struct with the current map and the new string as field). `Annotations` should always be all for the same query.

**Comment by zenador** [comment]
Now it looks like `PromQL warning: quantile value should be between 0 and 1 not 50.00 (at 19:21)`. Only implemented for the quantile warning as the rest aren't directly related to user input, only the returned metrics, so we display the relevant metric name instead. Do we need to pass in the full query string? It seems like that would require a lot of code changes, and the result is something we can already show (the `50`). I'm using `interface{}` for now as I'm not sure if we want to stick with this approach, and there's an import cycle if we use `parser.PositionRange`. If we stick with this, I could try moving `parser.PositionRange` into its own package or something.

**Comment by beorn7** [comment]
> Now it looks like PromQL warning: quantile value should be between 0 and 1, not 50.00 (at 19:21).

It would be cool to keep the syntax consistent with the full errors we already have. But that's just a minor cosmetic thing. See here:

image

The real problem is that the position is just "characters into the query", ignoring line breaks. The full errors tell you the line number and the character position within that line. There is logic for that in the `Error()` method of `ParseErr`, as I said [earlier](https://github.com/prometheus/prometheus/pull/12152#issuecomment-1690055651). See the code [here](https://github.com/prometheus/prometheus/blob/d73b4acb301df244eacb0d50189096abc177bd07/promql/parser/parse.go#L132-L152) (interestingly, it only points to the start of the character range in question, so we should do the same, which also means we only need the start position as a parameter for the annotations).

For this logic, we need the full query string, which answers your other question.

My wild idea also mentioned above was to let the engine add the full query string to the `Annotations` (making it a struct with the current map as a field and then a separate field for the query string). This could be done after the fact (i.e. just before the `Annotations` are returned to the caller). Not sure if that works, but it would not require us to propagate the query string everywhere.

So ideally, the last warning in the screenshot above should read something like: `PromQL warning: 1:20: quantile value should be between 0 and 1 but is 1.1` (BTW: Where is that weird trailing 0 coming from?).

> Only implemented for the quantile warning as the rest aren't directly related to user input, only the returned metrics, so we display the relevant metric name instead.

I think it's still very helpful to display the position. See the first warning in my screenshot above. Is it referring to the first or the second occurrence of `up`? It would be cool if it read: `PromQL warning: 1.25: no bucket label or malformed bucket label value on metric up`.

>  I'm using interface{} for now as I'm not sure if we want to stick with this approach, and there's an import cycle if we use parser.PositionRange. If we stick with this, I could try moving parser.PositionRange into its own package or something.

Yeah, I was afraid that would happen. Since the quite elaborate logic to convert the start position into an actual "line:pos" pair has to go somewhere where it is accessible to the PromQL engine, too, we need to move stuff around anyway. Maybe it could live in the promql package itself?

Having said all that: I do think we should provide the line:pos information, but we don't have to do it in the first iteration. It might be better to finish this PR up without it if it gets too involved. And the add this information later.

File: promql/parser/lex_test.go
Comment (beorn7) [inline promql/parser/lex_test.go:19]: This has to go into a separate block below, separated by a blank line from `"github.com/stretchr/testify/require"`.

**Comment by beorn7** [comment]
I guess the formatting with the position info at the end is generally OK. (I don't think that the formatting of the parser errors is optimal, anyway – so maybe that would be a separate task to rework all those error messages into something more readable and nicer to the eye.)

About the lint error: It's for file `promql/parser/lex_test.go`, which isn't one of the autogenerated ones (is it?). It should be easy to fix. I left a comment there.

I'll do a detailed code review ASAP.

**Comment by beorn7** [comment]
In the meantime, paging @roidelapluie again: It would be very good to get a general call about the changes in the API as discussed above and previously on [Slack](https://cloud-native.slack.com/archives/C01AUBA4PFE/p1692127732941169). Currently, we just add all the warnings and the one info annotation we have so far to the warnings. But there will be more "info" style annotations in the future (i.e. "warnings" that might be false positives and therefore should maybe go to their own info category). So:
- Should we do it as it is done here now (everything as a warning), at least for now? (Assuming that changing warnings isn't considered a breaking change.)
- Should we do it as it is done here now but hide the new annotations behind a feature flag so that we don't have to worry about a breaking change if we change our mind?
- Should we introduce a new "info" section in the API now (and adjust the UI accordingly to display them differently, in a green box or something)?

Maybe @juliusv could also chime in, especially if we do the latter and add something new to the UI.

**Comment by beorn7** [comment]
(Disclaimer: I still have to do the detailed review.)

I hijacked an unrelated meeting with @roidelapluie to find out what he thinks about this.

He brought up the concern that users that regularly create false positives for INFO level annotations might get rather annoyed by them (imagine someone using misnamed counters for historical reasons), so it would be good if INFO level annotations could be disabled (e.g. in the UI or even via config). On the other hand, we shouldn't jump to conclusions here. We can also try as is (i.e. have all the annotations as warnings for now) and only change if users are unhappy indeed. @roidelapluie agreed that moving an annotations from WARN to INFO isn't really a breaking change, so we could do it even without hiding everything behind a feature flag.

_But_ it's quite important to never spam the user with many annotations. Currently, this can easily happen if an annotation is spawned for every element in a huge metric vector. So we definitely need to put a limit in there. I guess the UI should never display more than around 10 annotations, followed by a message like "10,453 more PromQL warnings".

While this could be implemented just in the UI, I think we should even go further and limit it in the API itself. We are just talking about helpful annotations here, and that shouldn't OOM the receiver just because 10,000 annotations are included. So maybe let's implement that in the API itself to only show the first 10 (or so) warnings and infos, and then add one more new warning/info that just tells you how many more have been skipped.

**Comment by zenador** [comment]
@beorn7 Added the limit for the API

**Review by beorn7 — CHANGES_REQUESTED** [review]
Very nice. I think everything is now in place. I have a lot of comments, but they are all just touching documentation, minor tweaks and such. Nothing substantial.

Or let's say the only potentially substantial part is if you were planning to address the two `TODO(zenador): Issue warning when plumbing is in place.` items in `engine.go` or if you have postponed them deliberately.

File: util/annotations/annotations.go
Comment (zenador) [reply util/annotations/annotations.go:77]: Since Annotations is a map instead of a slice, I get an error when I try that.

File: util/annotations/annotations.go
Comment (zenador) [reply util/annotations/annotations.go]: I left it in for the cases of malformed label value (not missing), where there's an error with parsing it as a float.

File: util/annotations/annotations.go
Comment (zenador) [reply util/annotations/annotations.go]: This is confusing, I don't see any of those TODOs in my branch, I thought they were already addressed.

File: util/annotations/annotations.go
Comment (roidelapluie) [inline util/annotations/annotations.go]: is it only promql? this is a generic package

Comment (zenador) [reply util/annotations/annotations.go]: It can include storage as well, so this has already been changed to the generic "%d more annotations omitted" as per Bjorn's recommendation.

File: util/annotations/annotations.go
Comment (beorn7) [reply util/annotations/annotations.go:77]: Yeah, my bad...

**Review by beorn7 — APPROVED** [review]
Thank you very much. This is a huge accomplishment.

I'll merge this now, but there is one more thing to do:

Rule evaluations have ignored warnings so far, which was probably fine as they only came from remote read. Now that warnings AKA annotations are much more useful for recording rules, we need to log and count the annotations (and maybe even display them in the rules page of the web UI). The starting point is probably the [EngineQueryFunc](https://github.com/prometheus/prometheus/blob/156222cc509d2f234589ba0aac6ae8f673f54a85/rules/manager.go#L196), where we need to extract the annotations from the `res` and return it as an additional return argument.

**Comment by zenador** [comment]
Thank you for your thorough reviews and all your suggestions!

File: promql/functions.go
Comment (bboreham) [inline promql/functions.go:91]: I think you should add `metricName==""` as a case that is not warned.
This would happen for instance if the input to `rate` comes from a calculation, and not from a series.

Comment (juliusv) [reply promql/functions.go:91]: Yeah... but is there a valid scenario where that can happen? Rates should always be done on raw counters coming from the storage before any other operation is done on them... to pass in a derived expression into `rate()` you would have to do a subquery to turn it into a range vector, and I'm not sure there's ever a case where it makes sense to do that for counters before rate-ing them?

Comment (bboreham) [reply promql/functions.go:91]: second derivative?

Comment (juliusv) [reply promql/functions.go:91]: Hmpf. Ok fine :)

**Comment by beorn7** [comment]
We already have https://github.com/prometheus/prometheus/issues/12945 and https://github.com/prometheus/prometheus/pull/12922 . Please discuss there. (I'm pretty sure we should not warn on nameless metrics. Once we have stronger typing in PromQL, we can do it properly.)

**Pull request merged**
